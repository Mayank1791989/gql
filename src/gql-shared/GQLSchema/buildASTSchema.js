/* @flow */
/* eslint-disable no-use-before-define */
/**
 * NOTE: patched version
 */
import { type ObjMap } from 'graphql/jsutils/ObjMap';
import {
  type DocumentNode,
  type SchemaDefinitionNode,
  type TypeDefinitionNode,
  type DirectiveDefinitionNode,
  GraphQLSkipDirective,
  GraphQLIncludeDirective,
  GraphQLDeprecatedDirective,
  Kind,
  specifiedScalarTypes,
  type BuildSchemaOptions,
} from 'graphql';

import ASTDefinitionBuilder from './ASTDefinitionBuilder';
import extendSchema from './extendSchema';

import GQLSchema from './GQLSchema';
import { type GQLError, SEVERITY, newGQLError } from 'gql-shared/GQLError';
import invariant from 'invariant';

import { PLACEHOLDER_TYPES } from './PlaceholderTypes';

export default function buildASTSchema( // eslint-disable-line complexity
  ast: DocumentNode,
  options?: BuildSchemaOptions,
): { schema: GQLSchema, errors: Array<GQLError> } {
  if (!ast || ast.kind !== Kind.DOCUMENT) {
    throw new Error('Must provide a document ast.');
  }

  let schemaDef: ?SchemaDefinitionNode = null;
  const errors: Array<GQLError> = [];
  const typeDefs: Array<TypeDefinitionNode> = [];
  const nodeMap: ObjMap<TypeDefinitionNode> = Object.create(null);
  const directiveDefs: Array<DirectiveDefinitionNode> = [];
  const extendSchemaAST = {
    kind: 'Document',
    definitions: [],
  };

  for (let i = 0; i < ast.definitions.length; i += 1) {
    const def = ast.definitions[i];
    switch (def.kind) {
      case Kind.SCHEMA_DEFINITION:
        if (schemaDef) {
          errors.push(
            newGQLError(
              'Must provide only one schema definition.',
              [schemaDef, def],
              SEVERITY.error,
            ),
          );
        }
        schemaDef = def;
        break;
      case Kind.SCALAR_TYPE_DEFINITION:
      case Kind.OBJECT_TYPE_DEFINITION:
      case Kind.INTERFACE_TYPE_DEFINITION:
      case Kind.ENUM_TYPE_DEFINITION:
      case Kind.UNION_TYPE_DEFINITION:
      case Kind.INPUT_OBJECT_TYPE_DEFINITION: {
        const typeName = def.name.value;
        if (nodeMap[typeName]) {
          errors.push(
            newGQLError(
              `Type "${typeName} was defined more than once."`,
              [def],
              SEVERITY.error,
            ),
          );
        } else {
          typeDefs.push(def);
          nodeMap[typeName] = def;
        }
        break;
      }
      case Kind.DIRECTIVE_DEFINITION: {
        const directiveName = def.name.value;
        if (directiveDefs.find(dir => dir.name.value === directiveName)) {
          errors.push(
            newGQLError(
              `Directive "${directiveName}" already exists in the schema. It ` +
                'cannot be redefined.',
              [def],
              SEVERITY.error,
            ),
          );
        } else {
          directiveDefs.push(def);
        }
        break;
      }
      // Type extension
      case Kind.OBJECT_TYPE_EXTENSION:
      case Kind.SCALAR_TYPE_EXTENSION:
      case Kind.INTERFACE_TYPE_EXTENSION:
      case Kind.UNION_TYPE_EXTENSION:
      case Kind.ENUM_TYPE_EXTENSION:
      case Kind.INPUT_OBJECT_TYPE_EXTENSION: {
        extendSchemaAST.definitions.push(def);
        break;
      }
      default:
        // for some future added types
        errors.push(
          newGQLError(
            `The ${def.kind} kind is not yet supported.`,
            [def],
            SEVERITY.warn,
          ),
        );
        break;
    }
  }

  const operationTypes = schemaDef
    ? getOperationTypes(schemaDef)
    : {
        query: nodeMap.Query,
        mutation: nodeMap.Mutation,
        subscription: nodeMap.Subscription,
      };

  const definitionBuilder = new ASTDefinitionBuilder(
    nodeMap,
    options,
    (typeNode, ofType) => {
      const typeName = typeNode.name.value;
      errors.push(
        newGQLError(
          `Type "${typeName}" not found.`,
          [typeNode],
          SEVERITY.error,
        ),
      );

      switch (ofType) {
        case 'scalarType':
          return PLACEHOLDER_TYPES.scalarType(typeName, (typeNode: any));
        case 'inputType':
          return PLACEHOLDER_TYPES.inputType(typeName, (typeNode: any));
        case 'outputType':
          return PLACEHOLDER_TYPES.outputType(typeName, (typeNode: any));
        case 'objectType':
          return PLACEHOLDER_TYPES.objectType(typeName, (typeNode: any));
        case 'interfaceType':
          return PLACEHOLDER_TYPES.interfaceType(typeName, (typeNode: any));
        default:
          return invariant(false, 'unhandled case');
      }
    },
  );

  // types build
  const types = definitionBuilder.buildTypes(typeDefs, 'scalarType');

  // add core types
  types.push(...specifiedScalarTypes);

  // directives
  const directives = directiveDefs.map(def =>
    definitionBuilder.buildDirective(def),
  );

  if (!directives.some(directive => directive.name === 'skip')) {
    directives.push(GraphQLSkipDirective);
  }

  if (!directives.some(directive => directive.name === 'include')) {
    directives.push(GraphQLIncludeDirective);
  }

  if (!directives.some(directive => directive.name === 'deprecated')) {
    directives.push(GraphQLDeprecatedDirective);
  }

  const isExtendSchemaRequired = extendSchemaAST.definitions.length > 0;

  // Note: While this could make early assertions to get the correctly
  // typed values below, that would throw immediately while type system
  // validation with validateSchema() will produce more actionable results.
  const schema = new GQLSchema({
    query: operationTypes.query
      ? (definitionBuilder.buildType(operationTypes.query, 'outputType'): any)
      : null,
    mutation: operationTypes.mutation
      ? (definitionBuilder.buildType(
          operationTypes.mutation,
          'outputType',
        ): any)
      : null,
    subscription: operationTypes.subscription
      ? (definitionBuilder.buildType(
          operationTypes.subscription,
          'outputType',
        ): any)
      : null,
    types,
    directives,
    astNode: schemaDef,
    assumeValid: isExtendSchemaRequired, // if we are extending schema then do validation later
    commentDescriptions: options ? options.commentDescriptions : undefined,
    typeDependenciesMap: definitionBuilder.getTypeDependenciesMap(),
  });

  if (!isExtendSchemaRequired) {
    return { schema, errors: [...errors, ...schema._errors] };
  }

  // if extension present in ast then extendSchema
  const result = extendSchema(schema, extendSchemaAST, options);
  return {
    schema: result.schema,
    errors: [...errors, ...result.errors],
  };

  function getOperationTypes(schemaDefNode: SchemaDefinitionNode) {
    const opTypes = {};
    schemaDefNode.operationTypes.forEach(operationType => {
      const typeName = operationType.type.name.value;
      const { operation } = operationType;
      if (opTypes[operation]) {
        errors.push(
          newGQLError(
            `Must provide only one ${operation} type.`,
            [operationType],
            SEVERITY.error,
          ),
        );
        return;
      }

      if (!nodeMap[typeName]) {
        errors.push(
          newGQLError(
            `Specified ${operation} type "${typeName}" not found.`,
            [operationType.type],
            SEVERITY.error,
          ),
        );
        return;
      }

      opTypes[operation] = operationType.type;
    });
    return opTypes;
  }
}
