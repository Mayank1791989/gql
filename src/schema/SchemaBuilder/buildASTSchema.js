/* @flow */
/* eslint-disable no-use-before-define */
/**
 * NOTE: patched version
 */
import find from 'graphql/jsutils/find';
import invariant from 'graphql/jsutils/invariant';
import keyValMap from 'graphql/jsutils/keyValMap';
import { valueFromAST } from 'graphql/utilities/valueFromAST';
import { TokenKind } from 'graphql/language/lexer';
import { getArgumentValues } from 'graphql/execution/values';

import {
  LIST_TYPE,
  NON_NULL_TYPE,
  SCHEMA_DEFINITION,
  SCALAR_TYPE_DEFINITION,
  OBJECT_TYPE_DEFINITION,
  INTERFACE_TYPE_DEFINITION,
  ENUM_TYPE_DEFINITION,
  UNION_TYPE_DEFINITION,
  INPUT_OBJECT_TYPE_DEFINITION,
  DIRECTIVE_DEFINITION,
} from 'graphql/language/kinds';

import type {
  Location,
  ASTNode,
  DocumentNode,
  DirectiveNode,
  TypeNode,
  SchemaDefinitionNode,
  TypeDefinitionNode,
  ScalarTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  InputValueDefinitionNode,
  InterfaceTypeDefinitionNode,
  UnionTypeDefinitionNode,
  EnumTypeDefinitionNode,
  InputObjectTypeDefinitionNode,
  DirectiveDefinitionNode,
} from 'graphql/language/ast';

import GQLSchema from './GQLSchema';
import getNamedTypeNode from './getNamedTypeNode';

import {
  GQLString,
  GQLInt,
  GQLFloat,
  GQLBoolean,
  GQLID,
} from '../../shared/GQLTypes';

import {
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLNonNull,
  GraphQLList,
  isInputType,
  isOutputType,
} from 'graphql/type/definition';

import type { // eslint-disable-line no-duplicate-imports
  GraphQLType,
  GraphQLInputType,
  GraphQLOutputType,
} from 'graphql/type/definition';

import {
  GQLObjectType,
  GQLInputObjectType,
  GQLInterfaceType,
  GQLEnumType,
  GQLScalarType,
  GQLUnionType,
  GQLDirective,
  GQLSkipDirective,
  GQLIncludeDirective,
  GQLDeprecatedDirective,
} from '../../shared/GQLTypes';

import type { GQLNamedType, GQLType } from '../../shared/GQLTypes';

import type { // eslint-disable-line no-duplicate-imports
  DirectiveLocationEnum,
} from 'graphql/type/directives';

// import {
//   __Schema,
//   __Directive,
//   __DirectiveLocation,
//   __Type,
//   __Field,
//   __InputValue,
//   __EnumValue,
//   __TypeKind,
// } from 'graphql/type/introspection';

import { newGQLError, SEVERITY } from '../../shared/GQLError';
import type { GQLError } from '../../shared/GQLError';

import { PLACEHOLDER_TYPES } from './PlaceholderTypes';

function buildWrappedType(
  innerType: GraphQLType,
  inputTypeAST: TypeNode,
): GQLType {
  if (inputTypeAST.kind === LIST_TYPE) {
    return new GraphQLList((buildWrappedType(innerType, inputTypeAST.type): any));
  }
  if (inputTypeAST.kind === NON_NULL_TYPE) {
    const wrappedType = buildWrappedType(innerType, inputTypeAST.type);
    invariant(!(wrappedType instanceof GraphQLNonNull), 'No nesting nonnull.');
    return new GraphQLNonNull((wrappedType: any));
  }
  return (innerType: any);
}

export function buildASTSchema(ast: DocumentNode): { schema: GQLSchema, errors: Array<GQLError> } {
  let schemaDef: ?SchemaDefinitionNode;

  const errors: Array<GQLError> = [];
  const typeDefs: Array<TypeDefinitionNode> = [];
  const nodeMap: {[name: string]: TypeDefinitionNode} = Object.create(null);
  const nodeMapWithAllReferences: {[name: string]: Array<TypeDefinitionNode>} = Object.create(null);
  const directiveDefs: Array<DirectiveDefinitionNode> = [];

  for (let i = 0; i < ast.definitions.length; i += 1) {
    const d = ast.definitions[i];
    switch (d.kind) {
      case SCHEMA_DEFINITION:
        if (schemaDef) {
          errors.push(newGQLError(
            'Must provide only one schema definition.',
            [schemaDef, d],
            SEVERITY.error,
          ));
        }
        schemaDef = d;
        break;
      case SCALAR_TYPE_DEFINITION:
      case OBJECT_TYPE_DEFINITION:
      case INTERFACE_TYPE_DEFINITION:
      case ENUM_TYPE_DEFINITION:
      case UNION_TYPE_DEFINITION:
      case INPUT_OBJECT_TYPE_DEFINITION: {
        const name = d.name.value;
        if (!nodeMap[name]) {
          typeDefs.push(d);
          nodeMap[name] = d;
        }
        //  storing all reference to detect multiple defintition with same name
        if (!nodeMapWithAllReferences[name]) { nodeMapWithAllReferences[name] = []; }
        nodeMapWithAllReferences[name].push(d);
        break;
      }
      case DIRECTIVE_DEFINITION:
        directiveDefs.push(d);
        break;
      default:
    }
  }

  // error for multi same name typeDef
  Object.keys(nodeMapWithAllReferences).forEach((name) => {
    if (nodeMapWithAllReferences[name].length > 1) {
      errors.push(newGQLError(
        `Schema must contain unique named types but contains multiple types named "${name}".`,
        nodeMapWithAllReferences[name].map(typeDefAST => typeDefAST.name),
        SEVERITY.error,
      ));
    }
  });

  let queryTypeName;
  let mutationTypeName;
  let subscriptionTypeName;

  if (schemaDef) {
    schemaDef.operationTypes.forEach((operationType) => {
      const typeName = operationType.type.name.value;
      if (operationType.operation === 'query') {
        queryTypeName = typeName;
      } else if (operationType.operation === 'mutation') {
        mutationTypeName = typeName;
      } else if (operationType.operation === 'subscription') {
        subscriptionTypeName = typeName;
      }
    });
  } else {
    if (nodeMap.Query) {
      queryTypeName = 'Query';
    }
    if (nodeMap.Mutation) {
      mutationTypeName = 'Mutation';
    }
    if (nodeMap.Subscription) {
      subscriptionTypeName = 'Subscription';
    }
  }

  if (!queryTypeName) {
    errors.push(newGQLError(
      'Must provide schema definition with query type or a type named Query.',
      null,
      SEVERITY.error,
    ));
  }

  const innerTypeMap = {
    String: GQLString,
    Int: GQLInt,
    Float: GQLFloat,
    Boolean: GQLBoolean,
    ID: GQLID,
  };

  const types = typeDefs.map(def => typeDefNamed(def.name.value, def)).filter(Boolean);

  // Adding default types
  types.push(
    innerTypeMap.String,
    innerTypeMap.Int,
    innerTypeMap.Float,
    innerTypeMap.Boolean,
    innerTypeMap.ID,
  );

  // directives
  const directives = directiveDefs.map(getDirective);

  // If specified directives were not explicitly declared, add them.
  if (!directives.some(directive => directive.name === 'skip')) {
    directives.push(GQLSkipDirective);
  }

  if (!directives.some(directive => directive.name === 'include')) {
    directives.push(GQLIncludeDirective);
  }

  if (!directives.some(directive => directive.name === 'deprecated')) {
    directives.push(GQLDeprecatedDirective);
  }

  const schema = new GQLSchema({
    query: queryTypeName ? getObjectType(nodeMap[queryTypeName]) : null,
    mutation: mutationTypeName ? getObjectType(nodeMap[mutationTypeName]) : null,
    subscription: subscriptionTypeName ? getObjectType(nodeMap[subscriptionTypeName]) : null,
    types,
    directives,
    nodeMap,
  });

  return { schema, errors: [...errors, ...schema._errors] };

  function getDirective(directiveNode: DirectiveDefinitionNode): GQLDirective {
    return new GQLDirective(directiveNode, {
      name: directiveNode.name.value,
      description: getDescription(directiveNode),
      locations: directiveNode.locations.map(
        node => ((node.value: any): DirectiveLocationEnum),
      ),
      args: directiveNode.arguments && makeInputValues(directiveNode.arguments),
    });
  }

  function getObjectType(typeNode: TypeDefinitionNode): ?GQLObjectType {
    const type = typeDefNamed(typeNode.name.value, typeNode);

    if (!(type instanceof GraphQLObjectType)) {
      errors.push(newGQLError(
        'AST must provide object type.',
        [typeNode],
        SEVERITY.error,
      ));
    }

    return (type: any);
  }

  function typeDefNamed(typeName: string, node: ASTNode): ?GQLNamedType {
    if (innerTypeMap[typeName]) {
      if (innerTypeMap[typeName].dependents) {
        innerTypeMap[typeName].dependents.push(node);
      }
      return innerTypeMap[typeName];
    }

    if (!nodeMap[typeName]) {
      // NOT found
      errors.push(newGQLError(
        `Type "${typeName}" not found.`,
        [node],
        SEVERITY.error,
      ));
      return null;
    }

    const innerTypeDef = makeSchemaDef(nodeMap[typeName]);
    innerTypeDef.node = nodeMap[typeName]; // add location info
    innerTypeMap[typeName] = innerTypeDef;
    return innerTypeDef;
  }

  // function isProducedAlready(typeNode: TypeNode) {
  //   return Boolean(innerTypeMap[getNamedTypeNode(typeNode).name.value]);
  // }

  function produceType(typeNode: TypeNode, defaultValue: (name: string) => GraphQLType): GQLType {
    const namedTypeNode = getNamedTypeNode(typeNode);
    const typeName = namedTypeNode.name.value;
    const typeDef = typeDefNamed(typeName, namedTypeNode) || defaultValue(typeName);
    return buildWrappedType(typeDef, typeNode);
  }

  function produceInputType(typeNode: TypeNode): GraphQLInputType {
    const type = produceType(typeNode, PLACEHOLDER_TYPES.inputType);
    if (!isInputType((type: any))) {
      errors.push(newGQLError(
        'Expected Input type.',
        [getNamedTypeNode(typeNode)],
        SEVERITY.error,
      ));
    }
    return (type: any);
  }

  function produceOutputType(typeNode: TypeNode): GraphQLOutputType {
    const type = produceType(typeNode, PLACEHOLDER_TYPES.outputType);
    if (!isOutputType((type: any))) {
      errors.push(newGQLError(
        'Expected Output type.',
        [getNamedTypeNode(typeNode)],
        SEVERITY.error,
      ));
    }
    return (type: any);
  }

  function produceObjectType(typeNode: TypeNode): GQLObjectType {
    const type = produceType(typeNode, PLACEHOLDER_TYPES.objectType);
    if (!(type instanceof GraphQLObjectType)) {
      errors.push(newGQLError(
        'Expected Object type.',
        [getNamedTypeNode(typeNode)],
        SEVERITY.error,
      ));
    }
    return (type: any);
  }

  function produceInterfaceType(typeNode: TypeNode): GQLInterfaceType {
    const type = produceType(typeNode, PLACEHOLDER_TYPES.interfaceType);
    if (!(type instanceof GraphQLInterfaceType)) {
      errors.push(newGQLError(
        'Expected Interface type.',
        [getNamedTypeNode(typeNode)],
        SEVERITY.error,
      ));
    }
    return (type: any);
  }

  function makeSchemaDef(def) {
    if (!def) {
      throw new Error('def must be defined');
    }
    switch (def.kind) {
      case OBJECT_TYPE_DEFINITION:
        return makeTypeDef(def);
      case INTERFACE_TYPE_DEFINITION:
        return makeInterfaceDef(def);
      case ENUM_TYPE_DEFINITION:
        return makeEnumDef(def);
      case UNION_TYPE_DEFINITION:
        return makeUnionDef(def);
      case SCALAR_TYPE_DEFINITION:
        return makeScalarDef(def);
      case INPUT_OBJECT_TYPE_DEFINITION:
        return makeInputObjectDef(def);
      default:
        throw new Error(`Type kind "${def.kind}" not supported.`);
    }
  }

  function makeTypeDef(def: ObjectTypeDefinitionNode) {
    const typeName = def.name.value;
    return new GQLObjectType(def, {
      name: typeName,
      description: getDescription(def),
      fields: () => makeFieldDefMap(def),
      interfaces: () => makeImplementedInterfaces(def),
    });
  }

  function makeFieldDefMap(
    def: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
  ) {
    return keyValMap(
      def.fields,
      field => field.name.value,
      field => ({
        type: produceOutputType(field.type),
        node: field,
        description: getDescription(field),
        args: makeInputValues(field.arguments),
        deprecationReason: getDeprecationReason(field.directives),
      }),
    );
  }

  function makeImplementedInterfaces(def: ObjectTypeDefinitionNode) {
    return def.interfaces && def.interfaces.map(iface => produceInterfaceType(iface));
  }

  function makeInputValues(values: Array<InputValueDefinitionNode>) {
    return keyValMap(
      values,
      value => value.name.value,
      (value) => {
        const type = produceInputType(value.type);
        return {
          type,
          node: value,
          description: getDescription(value),
          defaultValue: valueFromAST(value.defaultValue, type),
        };
      },
    );
  }

  function makeInterfaceDef(def: InterfaceTypeDefinitionNode) {
    const typeName = def.name.value;
    return new GQLInterfaceType(def, {
      name: typeName,
      description: getDescription(def),
      fields: () => makeFieldDefMap(def),
      resolveType: cannotExecuteSchema,
    });
  }

  function makeEnumDef(def: EnumTypeDefinitionNode) {
    const enumType = new GQLEnumType(def, {
      name: def.name.value,
      description: getDescription(def),
      values: keyValMap(
        def.values,
        enumValue => enumValue.name.value,
        enumValue => ({
          description: getDescription(enumValue),
          deprecationReason: getDeprecationReason(enumValue.directives),
          node: enumValue,
        }),
      ),
    });

    return enumType;
  }

  function makeUnionDef(def: UnionTypeDefinitionNode) {
    return new GQLUnionType(def, {
      name: def.name.value,
      description: getDescription(def),
      types: def.types.map(t => produceObjectType(t)),
      resolveType: cannotExecuteSchema,
    });
  }

  function makeScalarDef(def: ScalarTypeDefinitionNode) {
    return new GQLScalarType(def, {
      name: def.name.value,
      description: getDescription(def),
      serialize: () => null,
      // Note: validation calls the parse functions to determine if a
      // literal value is correct. Returning null would cause use of custom
      // scalars to always fail validation. Returning false causes them to
      // always pass validation.
      parseValue: () => false,
      parseLiteral: () => false,
    });
  }

  function makeInputObjectDef(def: InputObjectTypeDefinitionNode) {
    return new GQLInputObjectType(def, {
      name: def.name.value,
      description: getDescription(def),
      fields: () => makeInputValues(def.fields),
    });
  }
}

function getDeprecationReason(directives: ?Array<DirectiveNode>): ?string {
  const deprecatedAST = directives && find(
    directives,
    directive => directive.name.value === GQLDeprecatedDirective.name,
  );
  if (!deprecatedAST) {
    return null;
  }
  const { reason } = getArgumentValues(
    GQLDeprecatedDirective,
    deprecatedAST,
  );
  return (reason: any);
}

/**
 * Given an ast node, returns its string description based on a contiguous
 * block full-line of comments preceding it.
 */
export function getDescription(node: { loc?: Location }): ?string {
  const loc = node.loc;
  if (!loc) { return null; }
  const comments = [];
  let minSpaces;
  let token = loc.startToken.prev;
  while (
    token &&
    token.kind === TokenKind.COMMENT &&
    token.next && token.prev &&
    token.line + 1 === token.next.line &&
    token.line !== token.prev.line
  ) {
    const value = String(token.value);
    const spaces = leadingSpaces(value);
    if (minSpaces === undefined || spaces < minSpaces) {
      minSpaces = spaces;
    }
    comments.push(value);
    token = token.prev;
  }
  return comments
    .reverse()
    .map(comment => comment.slice(minSpaces))
    .join('\n');
}

// Count the number of spaces on the starting side of a string.
function leadingSpaces(str) {
  let i = 0;
  for (; i < str.length; i += 1) {
    if (str[i] !== ' ') {
      break;
    }
  }
  return i;
}

function cannotExecuteSchema() {
  throw new Error(
    'Generated Schema cannot use Interface or Union types for execution.',
  );
}
