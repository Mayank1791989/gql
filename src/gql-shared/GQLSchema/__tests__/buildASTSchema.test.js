/* @flow */
import { parse } from 'graphql';
import buildASTSchema from '../buildASTSchema';

test('report multiple schema definition', () => {
  const { errors } = buildASTSchema(
    parse(`
      type Query {
        id: String!
      }
      schema {
        query: Query
      }
      schema {
        query: Query
      }
    `),
  );

  expect(errors).toMatchSnapshot();
});

test('report multiple type definition', () => {
  const { errors } = buildASTSchema(
    parse(`
      type Query {
        id: String!
      }
      schema {
        query: Query
      }

      type Query {
        id: Int!
      }
    `),
  );

  expect(errors).toMatchSnapshot();
});

test('report multiple directive definition', () => {
  const { errors } = buildASTSchema(
    parse(`
      type Query {
        id: String!
      }
      schema {
        query: Query
      }

      directive @test on FIELD
      directive @test on FIELD_DEFINITION
    `),
  );

  expect(errors).toMatchSnapshot();
});

describe('unknown types', () => {
  it('output field type unknown', () => {
    const { errors } = buildASTSchema(
      parse(`
        type Query {
          id: xString!
        }
        schema {
          query: Query
        }
      `),
    );

    expect(errors).toMatchSnapshot();
  });

  it('output field type unknown', () => {
    const { errors } = buildASTSchema(
      parse(`
        type Query {
          id: String!
        }
        schema {
          query: Query
        }

        input SomeInput {
          value: xValue!
        }
      `),
    );

    expect(errors).toMatchSnapshot();
  });

  it('interface type unknown', () => {
    const { errors } = buildASTSchema(
      parse(`
        type Query implements xNode {
          id: String!
        }
        schema {
          query: Query
        }
      `),
    );

    expect(errors).toMatchSnapshot();
  });

  it('union type unknown', () => {
    const { errors } = buildASTSchema(
      parse(`
        type Query {
          id: String!
        }
        schema {
          query: Query
        }

        union Test = Query | xQuery
      `),
    );

    expect(errors).toMatchSnapshot();
  });
});

describe('supports type extension', () => {
  it('report if extending unknown type', () => {
    const { errors } = buildASTSchema(
      parse(`
        type Query {
          id: String!
        }
        schema {
          query: Query
        }

        extend type xTest {
          name: String!
        }

        extend interface xNode {
          name: String!
        }
      `),
    );
    expect(errors).toMatchSnapshot();
  });

  it('reports unsupported type extensions', () => {
    const { errors } = buildASTSchema(
      parse(`
        type Query {
          id: String!
        }
        schema {
          query: Query
        }
        directive @test on SCALAR | UNION

        scalar Test
        extend scalar Test @test

        union UnionTest = Query | String
        extend union UnionTest @test

        enum EnumTest { x, y, z }
        extend enum EnumTest @test

        input InputTest {
          id: String!
        }
        extend input InputTest {
          name: String!
        }
      `),
    );
    expect(errors).toMatchSnapshot();
  });
});
