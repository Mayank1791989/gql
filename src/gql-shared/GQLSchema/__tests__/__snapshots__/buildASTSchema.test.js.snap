// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report multiple directive definition 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 7,
        "line": 10,
        "path": "GraphQL request",
      },
    ],
    "message": "Directive \\"test\\" already exists in the schema. It cannot be redefined.",
    "severity": "error",
  },
]
`;

exports[`report multiple schema definition 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 7,
        "line": 5,
        "path": "GraphQL request",
      },
      Object {
        "column": 7,
        "line": 8,
        "path": "GraphQL request",
      },
    ],
    "message": "Must provide only one schema definition.",
    "severity": "error",
  },
]
`;

exports[`report multiple type definition 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 7,
        "line": 9,
        "path": "GraphQL request",
      },
    ],
    "message": "Type \\"Query was defined more than once.\\"",
    "severity": "error",
  },
]
`;

exports[`supports type extension report if extending unknown type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 9,
        "path": "GraphQL request",
      },
    ],
    "message": "Cannot extend type \\"xTest\\" because it does not exist in the existing schema.",
    "severity": "error",
  },
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 13,
        "path": "GraphQL request",
      },
    ],
    "message": "Cannot extend type \\"xNode\\" because it does not exist in the existing schema.",
    "severity": "error",
  },
]
`;

exports[`supports type extension reports unsupported type extensions 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 11,
        "path": "GraphQL request",
      },
    ],
    "message": "The ScalarTypeExtension kind is not yet supported by extendSchema().",
    "severity": "warn",
  },
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 14,
        "path": "GraphQL request",
      },
    ],
    "message": "The UnionTypeExtension kind is not yet supported by extendSchema().",
    "severity": "warn",
  },
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 17,
        "path": "GraphQL request",
      },
    ],
    "message": "The EnumTypeExtension kind is not yet supported by extendSchema().",
    "severity": "warn",
  },
  Object {
    "locations": Array [
      Object {
        "column": 9,
        "line": 22,
        "path": "GraphQL request",
      },
    ],
    "message": "The InputObjectTypeExtension kind is not yet supported by extendSchema().",
    "severity": "warn",
  },
]
`;

exports[`unknown types interface type unknown 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 31,
        "line": 2,
        "path": "GraphQL request",
      },
    ],
    "message": "Type \\"xNode\\" not found.",
    "severity": "error",
  },
]
`;

exports[`unknown types output field type unknown 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 15,
        "line": 3,
        "path": "GraphQL request",
      },
    ],
    "message": "Type \\"xString\\" not found.",
    "severity": "error",
  },
]
`;

exports[`unknown types output field type unknown 2`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 18,
        "line": 10,
        "path": "GraphQL request",
      },
    ],
    "message": "Type \\"xValue\\" not found.",
    "severity": "error",
  },
]
`;

exports[`unknown types union type unknown 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 30,
        "line": 9,
        "path": "GraphQL request",
      },
    ],
    "message": "Type \\"xQuery\\" not found.",
    "severity": "error",
  },
]
`;
